File: openrndr-jvm/openrndr-adopted-artifex/src/main/java/io/lacuna/artifex/Region2.java
package io.lacuna.artifex;

import io.lacuna.artifex.Ring2.Result;
import io.lacuna.artifex.utils.EdgeList;
import io.lacuna.artifex.utils.regions.Clip;
import io.lacuna.artifex.utils.regions.Hulls;
import io.lacuna.artifex.utils.regions.Monotonic;
import io.lacuna.artifex.utils.regions.Triangles;
import io.lacuna.bifurcan.LinearList;
import io.lacuna.bifurcan.Lists;

import java.util.Arrays;
import java.util.Comparator;

/**
 * @author ztellman
 */
public class Region2 {

  public final Ring2[] rings;
  public final Box2 bounds;

  public Region2(Iterable<Ring2> rings) {
    this(LinearList.from(rings).toArray(Ring2[]::new));
  }

  public Region2(Ring2[] rings) {
    this.rings = rings.clone();
    Arrays.sort(this.rings, Comparator.comparingDouble(r -> r.area));

    this.bounds = Arrays.stream(this.rings)
      .map(r -> r.bounds)
      .reduce(Box2.EMPTY, Box2::union);
  }

  ///

  public Ring2[] rings() {
    return rings;
  }

  public static Region2 of(Ring2... rings) {
    return new Region2(Lists.from(rings));
  }

  public Box2 bounds() {
    return bounds;
  }

  public Result test(Vec2 p) {
    for (Ring2 r : rings) {
      Result result = r.test(p);
      if (result.inside) {
        return result.curve == null && r.isClockwise
          ? Result.OUTSIDE
          : result;
      }
    }

    return Result.OUTSIDE;
  }

  public boolean contains(Vec2 p) {
    return test(p).inside;
  }

  /// transforms and set operations

  public Region2 transform(Matrix3 m) {
    return new Region2(Arrays.stream(rings).map(r -> r.transform(m)).toArray(Ring2[]::new));
  }

  public Region2 intersection(Region2 region) {
    return Clip.intersection(this, region);
  }

  public Region2 union(Region2 region) {
    return Clip.union(this, region);
  }

  public Region2 difference(Region2 region) {
    return Clip.difference(this, region);
  }

}

File: openrndr-kartifex/src/commonMain/kotlin/Region2.kt
package org.openrndr.kartifex

import org.openrndr.kartifex.utils.regions.difference
import org.openrndr.kartifex.utils.regions.intersection
import org.openrndr.kartifex.utils.regions.union

/**
 * Represents a 2D region composed of multiple rings, where each ring defines a contour of the region.
 * Multiple rings can be used to create complex shapes, including those with holes.
 *
 * @constructor Creates a `Region2` object with the specified rings.
 * @property rings An array of `Ring2` objects that represent the contours of the region.
 * @property bounds The bounding box, represented by `Box2`, that encompasses all the rings in the region.
 */
class Region2(val rings: Array<Ring2>) {
    val bounds: Box2

    constructor(rings: Iterable<Ring2>) : this(
        rings.toList().toTypedArray()
    )

    fun test(p: Vec2): Ring2.Result {
        for (r in rings) {
            val result: Ring2.Result = r.test(p)
            if (result.inside) {
                return if (result.curve == null && r.isClockwise) Ring2.Result.OUTSIDE else result
            }
        }
        return Ring2.Result.OUTSIDE
    }

    operator fun contains(p: Vec2): Boolean {
        return test(p).inside
    }

    /// transforms and set operations
    fun transform(m: Matrix3): Region2 {
        return Region2(
            rings
                .map { r: Ring2 ->
                    r.transform(
                        m
                    )
                }.toTypedArray()
        )
    }

    fun intersection(region: Region2): Region2 {
        return intersection(this, region)
    }

    fun union(region: Region2): Region2 {
        return union(this, region)
    }

    fun difference(region: Region2): Region2 {
        return difference(this, region)
    }

    companion object {
        fun of(vararg rings: Ring2): Region2 {
            return Region2(rings.toList())
        }
    }

    init {
        rings.sortBy { it.area }

        bounds = rings.map { r: Ring2 -> r.bounds }
            .fold(Box2.EMPTY) { obj, b -> obj.union(b) }
    }
}
